src/main/flight/alt_hold.c

```diff
@@ -0,0 +1,241 @@
+}
+
+void simplePidInit(float kp, float ki, float kd)
+{
+    simplePid.kp = kp;
+    simplePid.ki = ki;
+    simplePid.kd = kd;
+    simplePid.lastAltitude = 0.0f;
+    simplePid.integral = 0.0f;
+}
+
+void altHoldReset(void)
+{
+    altHoldState.targetAltitudeCm = altHoldState.measuredAltitudeCm; // avoid jump at start
+    simplePid.lastAltitude = altHoldState.measuredAltitudeCm;        // avoid jump at start
+    simplePid.integral = 0.0f;
+    altHoldState.throttleOut = positionConfig()->hover_throttle;
+    altHoldState.targetAltitudeDelta = 0.0f;
+}
+
+void altHoldInit(void)
+{
+    simplePidInit(
+        ALT_HOLD_PID_P_GAIN * altholdConfig()->altHoldPidP,
+        ALT_HOLD_PID_I_GAIN * altholdConfig()->altHoldPidI,
+        ALT_HOLD_PID_D_GAIN * altholdConfig()->altHoldPidD);
+        // the multipliers are base scale factors
+        // iTerm is relatively weak, intended to be slow moving to adjust baseline errors
+        // the Hover value is important otherwise takes time for iTerm to correct
+        // High P will wobble readily
+        // with these scalers, the same numbers as for GPS Rescue work OK for altHold
+        // the goal is to share these gain factors, if practical for all altitude controllers
+
+    //setup altitude D filter
+    const float cutoffHz = 0.01f * positionConfig()->altitude_d_lpf; // default 1Hz, time constant about 160ms
+    const float gain = pt2FilterGain(cutoffHz, taskIntervalSeconds);
+    pt2FilterInit(&altHoldDeltaLpf, gain); 
+
+    altHoldState.isAltHoldActive = false;
+    altHoldReset();
+}
+
+void altHoldProcessTransitions(void) {
+
+    if (FLIGHT_MODE(ALTHOLD_MODE)) {
+        if (!altHoldState.isAltHoldActive && isAltitudeAvailable()) {
+            altHoldReset();
+            altHoldState.isAltHoldActive = true;
+        }
+    } else {
+        altHoldState.isAltHoldActive = false;
+    }
+
+    // ** the transition out of alt hold (exiting altHold) may be rough.  Some notes... **
+    // The original PR had a gradual transition from hold throttle to pilot control throttle
+    // using !(altHoldRequested && altHoldState.isAltHoldActive) to run an exit function
+    // a cross-fade factor was sent to mixer.c based on time since the flight mode request was terminated
+    // it was removed primarily to simplify this PR
+
+    // hence in this PR's the user's throttle needs to be close to the hover throttle value on exiting altHold
+    // its not so bad because the 'target adjustment' by throttle requires that
+    // user throttle must be not more than half way out from hover for a stable hold
+}
+
+void altHoldUpdateTargetAltitude(void)
+{
+    // The user van raise or lower the target altitude with throttle up;  there is a big deadband.
+    // Max rate for climb/descend is 1m/s by default (up to 2.5 is allowed but overshoots a fair bit)
+    // If set to zero, the throttle has no effect.
+
+    // Some people may not like throttle being able to change the target altitude, because:
```

Comment:
So that user has one switch to enable altitude hold, and another to either be stable, or climb or descend... yes that's possible.
However, now that I've flown it, there are a lot of advantages to providing target height adjustment by throttle.
In the case of a GPS, they can vary in altitude by some meters, especially when gaining or losing sats.  By having adjustment with throttle, these can be fixed while hovering.  
Also it is very helpful for adjusting the speed of a landing, or fine-tuning the hover point.
And since most pilots are automatic to adjust throttle if quad is too low, the throttle adjustment method is intuitive.
At this point I'd prefer to stick with the throttle adjustment method.